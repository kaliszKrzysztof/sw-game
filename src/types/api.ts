export type ApiCollectionSuccessResponse<T> = {
  count: number;
  next: string;
  previous: string | null;
  results: T[];
};

export type ApiErrorResponse = {
  status: number;
  message: string;
};
