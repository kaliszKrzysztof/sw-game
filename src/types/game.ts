export type Attribute = {
  name: string;
  value: string | number;
};

export interface GameItemData {
  title: string;
  attributes: Attribute[];
  power: number;
}
