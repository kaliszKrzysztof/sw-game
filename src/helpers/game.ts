import { GameItemData } from 'types/game';

type FightResult = {
  scores: [number, number];
  winner: string;
  draw?: boolean;
};

export const fight = (opponents: GameItemData[]): FightResult => {
  const [opponent1, opponent2] = opponents;
  if (opponent1.power > opponent2.power) {
    return {
      winner: opponent1.title,
      scores: [1, 0],
    };
  }
  if (opponent1.power < opponent2.power) {
    return {
      winner: opponent2.title,
      scores: [0, 1],
    };
  }
  return {
    winner: '',
    scores: [0, 0],
    draw: true,
  };
};
