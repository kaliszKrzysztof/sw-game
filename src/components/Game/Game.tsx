import React from 'react';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';
import { GameItemData } from 'types/game';
import { sleep } from 'helpers/app';
import GameCard from 'components/GameCard';
import Loader from 'components/Loader';
import { fight } from 'helpers/game';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    cardsWrapper: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      marginBottom: theme.spacing(3),
      [theme.breakpoints.up('sm')]: {
        flexDirection: 'row',
      },
    },
  }),
);

interface GameProps {
  playButtonText?: string;
  onPlayClick: () => [GameItemData, GameItemData];
}

export const getCardStatus = (
  title: string,
  winner: string,
  draw: boolean,
): 'draw' | 'win' | 'lose' => {
  if (draw) {
    return 'draw';
  }
  if (winner) {
    if (winner === title) {
      return 'win';
    }
    return 'lose';
  }
  return undefined;
};

const Game: React.FC<GameProps> = ({
  playButtonText = 'Play the game',
  onPlayClick,
}) => {
  const classes = useStyles();
  const [resources, setResources] = React.useState<
    ReturnType<typeof onPlayClick>
  >(null);
  const [scores, setScores] = React.useState([0, 0]);
  const [loading, setLoading] = React.useState(false);
  const [winner, setWinner] = React.useState('');
  const [draw, setDraw] = React.useState(false);

  const handlePlayClick = async (): Promise<void> => {
    setLoading(true);
    await sleep(1000);
    setWinner('');
    setDraw(false);
    const randomResources = onPlayClick();
    if (!randomResources) {
      setResources(null);
      return;
    }
    setResources(randomResources);
    await sleep(1000);
    const { winner: newWinner, scores: newScores, draw: newDraw } = fight(
      randomResources,
    );
    setWinner(newWinner);
    setDraw(newDraw);
    setScores(([prevFirstScore, prevSecondScore]) => [
      prevFirstScore + newScores[0],
      prevSecondScore + newScores[1],
    ]);
    setLoading(false);
  };

  return (
    <Box mt={4}>
      {loading && <Loader data-testid="loader" size={80} />}
      <Container maxWidth="md">
        <Box display="flex" justifyContent="space-between" mb={3}>
          <Typography variant="h4" data-testid="score1">
            {scores[0]}
          </Typography>
          <Typography variant="h4" data-testid="score2">
            {scores[1]}
          </Typography>
        </Box>
        {resources?.length > 0 && (
          <div className={classes.cardsWrapper}>
            {resources.map(({ title, attributes }, index) => (
              <React.Fragment key={title}>
                <Box flex={1} width="100%">
                  <GameCard
                    title={title}
                    attributes={attributes}
                    expanded
                    status={getCardStatus(title, winner, draw)}
                  />
                </Box>
                {index < resources.length - 1 && (
                  <Box alignSelf="center" p={4}>
                    VS
                  </Box>
                )}
              </React.Fragment>
            ))}
          </div>
        )}
        <Box mb={3}>
          <Button
            fullWidth
            data-testid="play"
            variant="contained"
            color="primary"
            size="large"
            onClick={handlePlayClick}
          >
            {playButtonText}
          </Button>
        </Box>
        <Box>
          <Button
            data-testid="reset"
            fullWidth
            variant="contained"
            onClick={() => setScores([0, 0])}
            size="large"
            disabled={scores.every((score) => score === 0)}
          >
            Reset scores
          </Button>
        </Box>
      </Container>
    </Box>
  );
};

export default Game;
