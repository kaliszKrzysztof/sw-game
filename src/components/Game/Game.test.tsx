import React from 'react';
import {
  fireEvent,
  render,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { GameItemData } from 'types/game';
import Game, { getCardStatus } from './Game';

const gameItem1: GameItemData = {
  title: 'item1',
  attributes: [],
  power: 100,
};

const gameItem2: GameItemData = {
  title: 'item2',
  attributes: [],
  power: 200,
};

const gameItem3: GameItemData = {
  title: 'item3',
  attributes: [],
  power: 100,
};

describe('<Game />', () => {
  it('should render', () => {
    render(<Game onPlayClick={() => null} />);
  });

  it('should match snapshot', () => {
    const { container } = render(<Game onPlayClick={() => null} />);
    expect(container).toMatchSnapshot();
  });

  it('should have correct initial state', () => {
    const { getByTestId, queryAllByTestId } = render(
      <Game onPlayClick={() => null} />,
    );
    const playButton = getByTestId('play');
    const resetButton = getByTestId('reset');
    const gameCards = queryAllByTestId('game-card');
    expect(playButton).not.toBeDisabled();
    expect(resetButton).toBeDisabled();
    expect(gameCards).toHaveLength(0);
  });

  it('should be able to play the game', async () => {
    jest.useFakeTimers();
    const onPlayClick = jest.fn();
    onPlayClick.mockReturnValueOnce([gameItem1, gameItem2]);
    const {
      getByTestId,
      getAllByTestId,
      queryAllByTestId,
      findAllByTestId,
    } = render(<Game onPlayClick={onPlayClick} />);
    const playButton = getByTestId('play');
    const gameCards = queryAllByTestId('game-card');
    expect(gameCards).toHaveLength(0);
    fireEvent.click(playButton);
    expect(getByTestId('loader')).toBeInTheDocument();
    expect(await findAllByTestId('game-card')).toHaveLength(2);
    await waitForElementToBeRemoved(() => getByTestId('loader'));

    // check for second element winner
    expect(getAllByTestId('game-card')[0]).toHaveAttribute(
      'style',
      'border-color: #f44336; color: rgb(244, 67, 54);',
    );
    expect(getAllByTestId('game-card')[1]).toHaveAttribute(
      'style',
      'border-color: #149414; color: rgb(20, 148, 20);',
    );
    expect(getByTestId('score1').textContent).toEqual('0');
    expect(getByTestId('score2').textContent).toEqual('1');

    // check for first element winner
    onPlayClick.mockReturnValueOnce([gameItem2, gameItem1]);
    fireEvent.click(playButton);
    expect(getByTestId('loader')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => getByTestId('loader'));

    expect(getAllByTestId('game-card')[0]).toHaveAttribute(
      'style',
      'border-color: #149414; color: rgb(20, 148, 20);',
    );
    expect(getAllByTestId('game-card')[1]).toHaveAttribute(
      'style',
      'border-color: #f44336; color: rgb(244, 67, 54);',
    );
    expect(getByTestId('score1').textContent).toEqual('1');
    expect(getByTestId('score2').textContent).toEqual('1');

    // check for draw
    onPlayClick.mockReturnValueOnce([gameItem3, gameItem1]);
    fireEvent.click(playButton);
    expect(getByTestId('loader')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => getByTestId('loader'));

    expect(getAllByTestId('game-card')[0]).toHaveAttribute(
      'style',
      'border-color: #fccf14; color: rgb(252, 207, 20);',
    );
    expect(getAllByTestId('game-card')[1]).toHaveAttribute(
      'style',
      'border-color: #fccf14; color: rgb(252, 207, 20);',
    );
    expect(getByTestId('score1').textContent).toEqual('1');
    expect(getByTestId('score2').textContent).toEqual('1');

    jest.useRealTimers();
  });

  it('can reset score', async () => {
    jest.useFakeTimers();
    const onPlayClick = jest.fn();
    onPlayClick.mockReturnValueOnce([gameItem1, gameItem2]);
    const { getByTestId } = render(<Game onPlayClick={onPlayClick} />);
    const playButton = getByTestId('play');
    const resetButton = getByTestId('reset');
    fireEvent.click(playButton);
    expect(getByTestId('loader')).toBeInTheDocument();
    await waitForElementToBeRemoved(() => getByTestId('loader'));

    expect(getByTestId('score1').textContent).toEqual('0');
    expect(getByTestId('score2').textContent).toEqual('1');

    fireEvent.click(resetButton);

    expect(getByTestId('score1').textContent).toEqual('0');
    expect(getByTestId('score2').textContent).toEqual('0');

    jest.useRealTimers();
  });

  it('should return correct status', () => {
    expect(getCardStatus('item1', 'item1', false)).toEqual('win');
    expect(getCardStatus('item1', '', true)).toEqual('draw');
    expect(getCardStatus('item1', 'item2', false)).toEqual('lose');
  });
});
