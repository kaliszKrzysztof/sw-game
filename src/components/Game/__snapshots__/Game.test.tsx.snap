// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game /> should match snapshot 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-7"
  >
    <div
      class="MuiContainer-root MuiContainer-maxWidthMd"
    >
      <div
        class="MuiBox-root MuiBox-root-8"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4"
          data-testid="score1"
        >
          0
        </h4>
        <h4
          class="MuiTypography-root MuiTypography-h4"
          data-testid="score2"
        >
          0
        </h4>
      </div>
      <div
        class="MuiBox-root MuiBox-root-9"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
          data-testid="play"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Play the game
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiBox-root MuiBox-root-10"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeLarge MuiButton-sizeLarge Mui-disabled MuiButton-fullWidth Mui-disabled"
          data-testid="reset"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Reset scores
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
